File: app/src/main/java/com/baijum/ukufretboard/ui/DailyChallengeView.kt
package com.baijum.ukufretboard.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.FilledTonalButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.baijum.ukufretboard.data.DailyChallengeGenerator

/**
 * Daily Challenges screen — shows today's 3 challenges.
 *
 * Each challenge suggests a specific activity (learn a chord, do a quiz,
 * practice a song, etc.) and provides a button to navigate to the relevant
 * section of the app.
 *
 * @param onNavigate Callback to navigate to a specific section index.
 */
@Composable
fun DailyChallengeView(
    onNavigate: (Int) -> Unit = {},
    modifier: Modifier = Modifier,
) {
    val challenges = remember { DailyChallengeGenerator.today() }

    Column(
        modifier = modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 16.dp, vertical = 8.dp),
    ) {
        // Header
        Row(
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                Icons.Filled.Star,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(28.dp),
            )
            Spacer(modifier = Modifier.width(8.dp))
            Column {
                Text(
                    text = "Daily Challenges",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.semantics { heading() },
                )
                Text(
                    text = "Complete today's challenges to build your skills!",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // Challenge cards
        challenges.forEachIndexed { index, challenge ->
            ChallengeCard(
                number = index + 1,
                challenge = challenge,
                onGo = {
                    challenge.navTarget?.let { onNavigate(it) }
                },
            )
            if (index < challenges.lastIndex) {
                Spacer(modifier = Modifier.height(12.dp))
            }
        }

        Spacer(modifier = Modifier.height(24.dp))

        // Tip of the day
        Card(
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(12.dp),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.surfaceVariant,
            ),
        ) {
            Column(modifier = Modifier.padding(16.dp)) {
                Text(
                    text = "Tip of the Day",
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Spacer(modifier = Modifier.height(4.dp))
                val tips = listOf(
                    "Even 5 minutes of daily practice is better than 1 hour once a week.",
                    "Use a metronome when practising chord changes — start slow and build up.",
                    "When learning a new chord, play each string individually to check clarity.",
                    "Try humming along while strumming to develop your musical ear.",
                    "Practice in front of a mirror to check your hand position.",
                    "Focus on smooth transitions rather than speed — speed comes with time.",
                    "Learn songs you love! Motivation makes practice enjoyable.",
                )
                val tipIndex = java.util.Calendar.getInstance().get(java.util.Calendar.DAY_OF_YEAR) % tips.size
                Text(
                    text = tips[tipIndex],
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }
    }
}

/**
 * A single challenge card with a number, title, description, and action button.
 */
@Composable
private fun ChallengeCard(
    number: Int,
    challenge: DailyChallengeGenerator.DailyChallenge,
    onGo: () -> Unit,
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.secondaryContainer,
        ),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Challenge number badge
            Card(
                shape = RoundedCornerShape(12.dp),
                colors = CardDefaults.cardColors(
                    containerColor = MaterialTheme.colorScheme.primary,
                ),
            ) {
                Text(
                    text = "$number",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onPrimary,
                    modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp),
                )
            }

            Spacer(modifier = Modifier.width(12.dp))

            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = challenge.title,
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onSecondaryContainer,
                )
                Spacer(modifier = Modifier.height(2.dp))
                Text(
                    text = challenge.description,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSecondaryContainer.copy(alpha = 0.8f),
                )
            }

            if (challenge.navTarget != null) {
                Spacer(modifier = Modifier.width(8.dp))
                FilledTonalButton(onClick = onGo) {
                    Text("Go")
                    Spacer(modifier = Modifier.width(4.dp))
                    Icon(
                        Icons.AutoMirrored.Filled.ArrowForward,
                        contentDescription = "Go",
                        modifier = Modifier.size(16.dp),
                    )
                }
            }
        }
    }
}

File: iosApp/UkuleleCompanion/Views/DailyChallengeView.swift
import SwiftUI
import shared

struct DailyChallengeView: View {
    @EnvironmentObject var learnVM: LearnViewModel
    @State private var challenges: [DailyChallengeGenerator.DailyChallenge] = []
    @State private var completedIndices: Set<Int> = []

    private static let tips = [
        "Even 5 minutes of daily practice is better than 1 hour once a week.",
        "Use a metronome when practising chord changes — start slow and build up.",
        "When learning a new chord, play each string individually to check clarity.",
        "Try humming along while strumming to develop your musical ear.",
        "Practice in front of a mirror to check your hand position.",
        "Focus on smooth transitions rather than speed — speed comes with time.",
        "Learn songs you love! Motivation makes practice enjoyable."
    ]

    private var tipOfTheDay: String {
        let dayOfYear = Calendar.current.ordinality(of: .day, in: .year, for: Date()) ?? 1
        return Self.tips[dayOfYear % Self.tips.count]
    }

    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                if challenges.isEmpty {
                    Text("Loading challenges...")
                        .foregroundStyle(.secondary)
                } else {
                    ForEach(Array(challenges.enumerated()), id: \.offset) { index, challenge in
                        challengeCard(challenge, index: index)
                    }
                }

                tipCard
            }
            .padding()
        }
        .navigationTitle("Daily Challenge")
        .onAppear {
            loadChallenges()
        }
    }

    private var tipCard: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Image(systemName: "lightbulb.fill")
                    .foregroundStyle(.yellow)
                    .font(.title2)
                Text("Tip of the Day")
                    .font(.headline)
                    .accessibilityAddTraits(.isHeader)
            }
            Text(tipOfTheDay)
                .font(.subheadline)
                .foregroundStyle(.secondary)
        }
        .padding()
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color(.systemGray6))
        .cornerRadius(12)
    }

    private func challengeCard(_ challenge: DailyChallengeGenerator.DailyChallenge, index: Int) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Image(systemName: iconForType(challenge.type))
                    .font(.title2)
                    .foregroundStyle(Color.accentColor)
                VStack(alignment: .leading) {
                    Text(challenge.title)
                        .font(.headline)
                    Text(challenge.type.label)
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }
                Spacer()
                if completedIndices.contains(index) {
                    Image(systemName: "checkmark.circle.fill")
                        .foregroundStyle(.green)
                        .font(.title2)
                }
            }
            .accessibilityCombined(label: "\(challenge.title), \(challenge.type.label)", value: completedIndices.contains(index) ? "completed" : "not completed")

            Text(challenge.description_)
                .font(.subheadline)
                .foregroundStyle(.secondary)

            if !completedIndices.contains(index) {
                Button("Mark Complete") {
                    completedIndices.insert(index)
                    saveDayProgress(index: index)
                    learnVM.recordActivity()
                }
                .buttonStyle(.bordered)
                .frame(maxWidth: .infinity, alignment: .trailing)
            }
        }
        .padding()
        .background(Color(.systemGray6))
        .cornerRadius(12)
    }

    private func iconForType(_ type: DailyChallengeGenerator.ChallengeType) -> String {
        switch type {
        case .learnChord: return "music.note"
        case .theoryQuiz: return "questionmark.circle"
        case .practiceSong: return "music.note.list"
        case .scalePractice: return "music.quarternote.3"
        case .chordTransition: return "arrow.left.arrow.right"
        case .earTraining: return "ear"
        default: return "star"
        }
    }

    private func loadChallenges() {
        let list = DailyChallengeGenerator.shared.today()
        challenges = list.asArray(of: DailyChallengeGenerator.DailyChallenge.self)
        loadDayProgress()
    }

    private var dayKey: String {
        let cal = Calendar.current
        let year = cal.component(.year, from: Date())
        let day = cal.ordinality(of: .day, in: .year, for: Date()) ?? 1
        return "daily_\(year)_\(day)"
    }

    private func saveDayProgress(index: Int) {
        UserDefaults.standard.set(true, forKey: "\(dayKey)_\(index)")
    }

    private func loadDayProgress() {
        completedIndices = []
        for i in 0..<3 {
            if UserDefaults.standard.bool(forKey: "\(dayKey)_\(i)") {
                completedIndices.insert(i)
            }
        }
    }
}
